Add 3d thermal forcing to OCN-GLC coupling

matthewhoffman opened the pull request:
This pull request extends the existing thermal-forcing coupling functionality from a single depth to a predefined set of z-levels. The implementation follows that of multiple elevation classes in the LND-GLC coupling. A new xml variable is added, GLC_NZOC. It has a default value of 0, meaning TF coupling is disabled. The standard value for when the coupling is activated is 30, which follows the TF z-levels used by ISMIP6. Other values are supported, and an additional example with 4 levels is included. A new coupler module, glc_zocnclass_mod.F90, is added which is follows very closely from glc_elevclass_mod.F90. It contains subroutines and functions to initialize the z-level classes and to access the z-level values. It also has subroutines to provide the z-level indices as strings, which the coupling field initialization code and the component import/export routines uses to create GLC_NZOC fields, one for each layer.

This 3d TF coupling is used for ice-sheet outlet glacier facemelting (melting at vertical glacier margins terminating in the ocean), and so is used for a process that we do not anticipate resolving explicitly in MPAS-Ocean (or Omega). As such, this coupling should be included anytime both OCN and GLC are active, and logic has been added so that this is the case based on compset definition. There are versions of the GLC_NZOC variable in both the MPAS-Ocean and MALI Registries so that it can be set by namelist settings controlled by CIME/build operations.

The MPAS-Ocean mpas_ocn_time_average_coupled.F module has been modified to loop over the z-levels and calculate TF at each one. Then, the time-averaged 3d TF values are passed to the coupler, where they are remapped to the GLC mesh and passed to MALI. MALI assigns the values to its existing 3d TF field, and the uses its bathymetry-aware extrapolation routine to extrapolate the TF values from the edge of the MPAS-Ocean domain to wherever the ice-sheet margin is on the MALI mesh. From there, the TF is used to force the facemelting parameterization.  Mapping files have been updated to not include extrapolation in the mapping so as to allow the bathymetry-aware extrapolation in MALI to operate, where appropriate.

The facemelting flux in MALI is exported to the previously unused Fogg_rofi flux, which already gets combined in the coupler with ice runoff to be imported in MPAS-Ocean in the Foxx_rofi flux. Long term, we may want to separate these runoff fluxes so that the horizontal and vertical distribution of facemelting can be handled differently than other solid runoff, but this provides a reasonable first approximation.

Note that this 3d TF can also be used to force the ISMIP6 ice-shelf basal melt parameterization in MALI, which provides a less sophisticated alternative to using ice-shelf basal melt rates calculated in the coupler (or MPAS-Ocean). This simpler approach will be used for the few, small ice shelves in Greenland, which are not included in the MPAS-Ocean meshes (and would not be resolved even if they were). This approach can also be used for Antarctica as an alternative that avoids grounding line migration issues in MPAS-Ocean by taking advantage of the TF extrapolation in MALI.  These different methods for ice-shelf basal melt fluxes are handled through a new XML variable OCN_GLC_ISMF_COUPLING, which specifies how ice-shelf melt fluxes are handled in both MPAS-Ocean and MALI.

The previous 2d thermal forcing coupling has been removed.  Tests have been updated to test this new 3d TF capability for grids with both GIS and AIS.  

[NML]
[non-BFB] only for configurations with active MALI and ocn

matthewhoffman commented:
Discussion and testing documented at https://github.com/E3SM-Ocean-Discussion/E3SM/pull/121

xylar commented:
Assuming https://github.com/E3SM-Project/E3SM/pull/7195 goes in before this, there are a few things we'll need to fix after rebasing on to that merge:
* fix one more `ocn_ismf`
* update the mapping files between ocn and glc to following this conversion.

jonbob commented:
@rljacob -- we've asked you to review this as well since it touches the coupler

matthewhoffman commented:
@rljacob , I'd be happy to meet and walk you through it if that would be easier for you.

matthewhoffman commented:
FYI, I plan to rebase this deal with known conflicts with https://github.com/E3SM-Project/E3SM/pull/7195, now that 7195 is on master. I hope to do that tomorrow.

xylar commented:
I'm running a `GMPAS-JRA1p5-DIB-PISMF-SIS.TL319_oQU240wLI_ais8to30` case and seeing:
```
 84:  0 0x0000000000012cf0 __funlockfile()  :0
 84:  1 0x0000000001e787bb glc_comp_mct_mp_init_ocean_z_levels_.A()  /lcrc/group/e3sm/ac.xylar/scratch/chrys/20250410.GMPAS-JRA1p5-DIB-PISMF-SIS.TL319_oQU240wLI_ais8to30.chrysalis/mpas-albany-landice/driver/glc_comp_mct.f90:1829
 84:  2 0x0000000001e7c376 glc_comp_mct_mp_glc_init_mct_()  /lcrc/group/e3sm/ac.xylar/scratch/chrys/20250410.GMPAS-JRA1p5-DIB-PISMF-SIS.TL319_oQU240wLI_ais8to30.chrysalis/mpas-albany-landice/driver/glc_comp_mct.f90:706
 84:  3 0x0000000000452c04 component_mod_mp_component_init_cc_()  /gpfs/fs1/home/ac.xylar/e3sm_work/E3SM/matthewhoffman/ocn-glc/3d-tf-ocn-glc/driver-mct/main/component_mod.F90:257
 84:  4 0x000000000043f1dc cime_comp_mod_mp_cime_init_()  /gpfs/fs1/home/ac.xylar/e3sm_work/E3SM/matthewhoffman/ocn-glc/3d-tf-ocn-glc/driver-mct/main/cime_comp_mod.F90:1531
 84:  5 0x000000000044f89a MAIN__()  /gpfs/fs1/home/ac.xylar/e3sm_work/E3SM/matthewhoffman/ocn-glc/3d-tf-ocn-glc/driver-mct/main/cime_driver.F90:122
 84:  6 0x000000000041b9a2 main()  ???:0
 84:  7 0x000000000003ad85 __libc_start_main()  ???:0
 84:  8 0x000000000041b8ae _start()  ???:0
 84: =================================
```
I'm looking into this but wanted to give you a heads up that a fix is likely needed.

xylar commented:
@matthewhoffman, I think maybe you'd better take this one on.  It seems like maybe the `ismip6Melt` package isn't active but the code things there are 30 vertical z-levels and is trying to print them out even though the fields aren't allocated.

https://github.com/E3SM-Project/E3SM/blob/matthewhoffman/ocn-glc/3d-tf-ocn-glc/components/mpas-albany-landice/src/mode_forward/mpas_li_core_interface.F#L170-L179

https://github.com/E3SM-Project/E3SM/blob/matthewhoffman/ocn-glc/3d-tf-ocn-glc/components/mpas-albany-landice/driver/glc_comp_mct.F#L1832-L1834

matthewhoffman commented:
@xylar , I've rebased and made some changes:
* dealt with conflicts or issues needing updating during the rebase
* updated MALI grid files to fix the dimension issue we diagnosed (and other minor update Trevor found)
* fixed an out-of-bounds error that the other fixes exposed

I also touched base with Jon about needing some new mapping files for SOwISC12to30E3r4 to use TF coupling, now that that came in as part of the rebase.

I confirmed the following tests pass after the changes that could affect these grids:
* SMS_D_Ld5.TL319_oQU240wLI_gis4to40.MPAS_FOLISIO_JRA1p5.chrysalis_gnu
* SMS_D_Ld5.TL319_oQU240wLI_ais8to30.MPAS_FOLISIO_JRA1p5.chrysalis_gnu
* SMS_D_Ld5.TL319_IcoswISC30E3r5_ais4to20.MPAS_FOLISIO_JRA1p5.chrysalis_gnu 

However, SMS_D_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-SIS.chrysalis_gnu  still fails. The errors Xylar was seeing are now gone and the code gets through init for all components.  However, it dies with this error in the e3sm log file:
```
 55: MCT::m_AttrVect::indexRA_:: FATAL--attribute not found: "norm8wt" Traceback:
 55: |X|MCT::m_AttrVect::indexRA_
 55: 037.MCT(MPEU)::die.: from MCT::m_AttrVect::indexRA_()
```
I will need to look into this further, but am stopping here for now.

xylar commented on components/mpas-ocean/cime_config/buildnml:
```diff
@@ -429,7 +430,7 @@ def buildnml(case, caseroot, compname):
             logger.warning("WARNING: The specified compset is requesting ocean ICs spunup from a G-case")
             logger.warning("         But no file available for this grid.")
-        if ocn_ismf == 'data':
+        if ocn_glc_ismf_coupling == 'data_mpaso':
```
Great, glad you caught this one, too.  I needed this change in my test-merge branch.

matthewhoffman commented:
@xylar , I've pushed commits that will handle the situation where both OCN and GLC are active but we don't want TF coupling (e.g. MALI DATA and STATIC modes).  With these changes, `SMS_D_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-SIS.chrysalis_gnu` now runs without error.

matthewhoffman commented:
...and the compsets that do include TF coupling are now broken.  I'll wait to update again until I can get all test cases working again.

matthewhoffman commented:
I fixed a bug introduced in the previous commit and reran these tests to confirm things are working as intended:
* SMS_D_Ld5.TL319_oQU240wLI_gis4to40.MPAS_FOLISIO_JRA1p5.chrysalis_gnu PASS
* SMS_D_Ld5.TL319_oQU240wLI_ais8to30.MPAS_FOLISIO_JRA1p5.chrysalis_gnu PASS
* SMS_D_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-SIS.chrysalis_gnu PASS
* SMS_D_Ld5.TL319_IcoswISC30E3r5_ais4to20.MPAS_FOLISIO_JRA1p5.chrysalis_gnu PASS

Other than the needed mapping files for `SOwISC12to30E3r4`, everything is resolved that I am aware of.

matthewhoffman commented:
@jonbob , I added a commit that updates the SOwISC12to30E3r4 mapping files.  I successfully ran a test with `SMS_D_Ld5.TL319_SOwISC12to30E3r4_ais4to20.MPAS_FOLISIO_JRA1p5.chrysalis_gnu` to ensure that they don't throw any errors.

There are no remaining revisions to make that I am aware of.  @rljacob , this is at a good point for you to review if you would like to review it.

rljacob commented on cime_config/tests.py:
```diff
@@ -299,7 +299,10 @@
             "ERS_Ld5_D.T62_oQU240.GMPAS-IAF.mpaso-conservation_check",
             "ERS_Ld5_PS.ne30pg2_r05_IcoswISC30E3r5.CRYO1850-DISMF.mpaso-scaled_dib_dismf",
-            "ERS_Ld5.TL319_oQU240wLI_gis20.MPAS_LISIO_JRA1p5.mpaso-ocn_glc_tf_coupling",
+            # OCN/GLC 3d TF coupling GIS test:
```
should we run this e3sm_ocnice_stealth_features suite somewhere on a regular basis?

jonbob replied:
@rljacob -- it runs every night on gcp12

rljacob replied:
Oh I see.  Its included in extra coverage.

rljacob commented on cime_config/config_grids.xml:
```diff
@@ -6040,7 +6040,7 @@
       <map name="OCN2GLC_SHELF_FMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis20km_esmfaave.20240919.nc</map>
       <map name="OCN2GLC_SHELF_SMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis20km_esmfbilin.20240919.nc</map>
-      <map name="OCN2GLC_TF_SMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis20km_deeperThan300m.esmfneareststod.20240919.nc</map>
+      <map name="OCN2GLC_TF_SMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis20km_esmfbilin.20240919.nc</map>
```
Did tempestremap not work?

jonbob replied:
I still can't get it to work when both grids are non-global like this. I understand that mbtempest can do that, but I haven't really entrained that into my workflow

matthewhoffman commented:
@rljacob ,  I wanted to check-in to see if you want to review this further and where it sits on your priority list.  (We have some follow-up activities that are starting to stack up behind this.)

rljacob commented on cime_config/config_grids.xml:
```diff
@@ -6107,25 +6107,31 @@
       <map name="OCN2GLC_SHELF_SMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis4to40_esmfbilin.20250303.nc</map>
-      <map name="OCN2GLC_TF_SMAPNAME">cpl/gridmaps/oQU240wLI/map_oQU240wLI_to_gis4to40_esmfaave.20250303.nc</map>
+      <!-- due to a known issue with ocean extrapolation in MALI, as of 2025-04-04 it is necessary to use a mapping file with
+           extrapolation included for OCN2GLC_TF_SMAPNAME.  After that is fixed, this file should be switched to the
```
What does it mean to include extrapolation in the map?

matthewhoffman replied:
Including extrapolation in the map is done with the `--extrap_method nearestidavg` argument to ESMF_RegridWeightGen/ncremap.  

Excluding extrapolation is done with `--extrap_method none`.

rljacob commented on cime_config/config_grids.xml:
```diff
@@ -6107,25 +6107,31 @@
       <map name="GLC2OCN_SMAPNAME">cpl/gridmaps/mpas.gis4to40km/map_gis4to40_to_oQU240wLI_esmfaave.20250303.nc</map>
       <map name="GLC2OCN_LIQ_RMAPNAME">cpl/gridmaps/mpas.gis4to40km/map_gis4to40_to_oQU240wLI_nn.20250303.nc</map>
-      <map name="GLC2OCN_ICE_RMAPNAME">cpl/gridmaps/mpas.gis4to40km/map_gis4to40_to_oQU240wLI_nn.20250303.nc</map>
+      <map name="GLC2OCN_ICE_RMAPNAME">cpl/gridmaps/mpas.gis4to40km/map_gis4to40_to_oQU240wLI-nomask_nn.20250403.nc</map>
```
Instead of "nn" should it be "cstmnn" because it was made with a custom tool?

matthewhoffman replied:
Seems reasonable to me.  @jonbob , what do you think?

jonbob replied:
I think that's OK -- it's not generated by the smoothing code where we have generally added that label, as well as the amount of smoothing. But those files were made from an old cess code that was modified to work for us. When we make both files, smoothed and the nearest-neighbor one the smoothing is based on, we've historically left one as "nn" and the other as "rXXXeXXX.cstmnn". But I'm happy to change that if it's preferable

rljacob replied:
Its fine.  Is this code in a repo somewhere?

jonbob replied:
It's a modified version of a cesm/cime tool, but we're working on replacing it instead of tracking it. I'm happy to change the mapping filenames if that better fits the naming convention

jonbob replied:
@rljacob -- I think the names of these files is the only thing holding up this PR. If you prefer "cstmnn" I'm happy to change them

rljacob replied:
No you can leave the names as is.

rljacob reviewed (approved):
I'm surprised how few mods there were in the driver code.

matthewhoffman commented:
@rljacob , thanks for looking at this again.  There are so few changes to the driver code, because all the hard work was done in #6632.  Because this PR was just extending things from a single field to the same information flow at multiple depth levels, there was not a lot of driver code needing modification.

matthewhoffman commented:
Force-pushed after rebase to deal with known conflicts from #7210 being merged to master.

jonbob commented on cime_config/tests.py:
```diff
@@ -300,7 +300,10 @@
             "ERS_Ld5_PS.ne30pg2_r05_IcoswISC30E3r5.CRYO1850-DISMF.mpaso-scaled_dib_dismf",
-            "ERS_Ld5.TL319_oQU240wLI_gis20.MPAS_LISIO_JRA1p5.mpaso-ocn_glc_tf_coupling",
+            # OCN/GLC 3d TF coupling GIS test:
+            "ERS_Ld5.TL319_oQU240wLI_gis4to40.MPAS_FOLISIO_JRA1p5.mpaso-jra_1958",
```
@matthewhoffman -- it doesn't look like this stealth test is actually turning the feature on?

jonbob commented:
Testing shows expected results:
* NML DIFF only for SMS_D_Ld1.ne30pg2_r05_IcoswISC30E3r5.WCYCL1850.chrysalis_intel.allactive-wcprod
* NML and baseline DIFFs for e3sm_landice_developer using chrysalis_gnu

Merged to next

xylar commented:
Thanks @jonbob!!  And thanks @matthewhoffman for the hard work on this!
